File: src/main/java/kvdb/oo/v1/KVDB_V1.java
package kvdb.oo.v1;

import java.util.HashMap;
import java.util.Map;

public class KVDB_V1 {
    public static void main(String[] args) {

        Map<String, String> store = new HashMap<>();

        String value = new SetCommand("key", "value").execute(store);
        assert value.equals("value");

        value = new GetCommand("key").execute(store);
        assert value.equals("value");

        new DelCommand("key").execute(store);

        value = new GetCommand("key").execute(store);
        assert value == null;
    }
}

class GetCommand implements Command {
    private final String key;
    public GetCommand(String key) {
        this.key = key;
    }
    @Override
    public String execute(Map<String, String> store) {
        return store.get(this.key);
    }
}

class SetCommand implements Command {

    private final String key;
    private final String value;
    public SetCommand(String key, String value) {
        this.key = key;
        this.value = value;
    }
    @Override
    public String execute(Map<String, String> store) {
        store.put(this.key, this.value);
        return this.value;
    }
}

class DelCommand implements Command {
    private final String key;
    public DelCommand(String key) {
        this.key = key;
    }
    @Override
    public String execute(Map<String, String> store) {
        return store.remove(this.key);
    }
}

File: src/main/scala/kvdb/fp/v1/KVDB_V1.scala
package kvdb.fp.v1

import scala.collection.mutable.{HashMap => MutableHashMap}

object KVDB_V1 extends App {
  override def main(args: Array[String]): Unit = {
    // OO style invocation
    assert(Store.execute(SetCommand("key", "value")) == "value")
    assert(Store.execute(GetCommand("key")) == "value")
    Store.execute(DelCommand("key"))
    assert(Store.execute(GetCommand("key")) == null)

    // function invocation
    assert(execute(SetCommand("key", "value")) == "value")
    assert(execute(GetCommand("key")) == "value")
    execute(DelCommand("key"))
    assert(execute(GetCommand("key")) == null)
  }
}

trait Command
case class SetCommand(key: String, value: String) extends Command
case class GetCommand(key: String) extends Command
case class DelCommand(key: String) extends Command

object Store {

  val store = new MutableHashMap[String, String]()

  def execute(c: Command): String = {
    c match {
      case SetCommand(k, v) => store.put(k, v); v
      case GetCommand(k) => store.getOrElse(k, null)
      case DelCommand(k) => store.remove(k).orNull
    }
  }
}

// this is scala's way of defining a function as an object but callable as a first class function
object execute extends (Command => String) {
  override def apply(c: Command): String = Store.execute(c)
}